### PR TITLE
[SILOptimizer] Don't diagnose infinite recursion if a branch terminates the program

### PR DESCRIPTION
**Pull request by harlanhaskins**
(Re-applies commit e94450e, but alters the strategy)

This patch augments the infinite recursion checker to not warn if a
branch terminates, but still warns if a branch calls into something with
@_semantics("programtermination_point"). This way, calling fatalError
doesn't disqualify you for the diagnostic, but calling exit does.

This also removes the warning workaround in the standard library, and
annotates the internal _assertionFailure functions as
programtermination_points, so they get this treatment too.

rdar://45080912

### CODE REVIEW

File: lib/SIL/SILInstructions.cpp [hunk 1]
```diff
@@ -1061,6 +1061,30 @@ bool TermInst::isFunctionExiting() const {
   llvm_unreachable("Unhandled TermKind in switch.");
 }
 
+bool TermInst::isProgramTerminating() const {
+  switch (getTermKind()) {
+    case TermKind::BranchInst:
```
Comment (gottesmm) [inline lib/SIL/SILInstructions.cpp]: Did you run this through git-clang-format?

**Review by gottesmm — COMMENTED** [review]
Looks great. Some questions:

1. Did you use clang-format? I thought that clang-format formatted the case statements so that they are directly under the switch?
2. I was thinking about the problem that was run into. Is it possible for us to get tests for that. I am hesitant if it is truly needed since the implementation is so simple.

But functionally, LGTM! I really like where we iterated to with this. = ).

File: lib/SIL/SILInstructions.cpp (hunk 1 above)
Comment (harlanhaskins) [reply lib/SIL/SILInstructions.cpp]: Ah! Thanks, I will. Xcode's reindent-on-paste indents each case 😢

**Comment by harlanhaskins** [comment]
@gottesmm For posterity, the problem that caused this to revert was that `_emptyStringAddressBits` could possibly have side effects, but it's only used on i386/arm, which I didn't test when I merged the first PR. Since that function is called when creating the empty string, the call to `fatalError()` implicitly called it after inlining the default argument initializer. The presence of something with side effects defeated the ARC inertness check, which it turns out was the wrong check for infinite recursion detection anyway.

By checking explicitly for the presence of an apply of something with `@_semantics("programtermination_point")`, we disregard inertness when looking for infinite recursion.

All that to say that a possible test to ensure we don't regress on this specific bug is:

```swift
func mayHaveSideEffects() {}

func x() { // expected-warning {{all paths through this function will call itself}}
  if Bool.random() {
    // execute some side effects
    mayHaveSideEffects()
    fatalError()
  }
  x()
}
```

And I'll add that test.

**Comment by harlanhaskins** [comment]
@swift-ci please test

**Comment by harlanhaskins** [comment]
rdar://45080912

**Pull request merged**